lib/models/plugin-registry.js

```diff
@@ -0,0 +1,278 @@
+        config = self._mergeAuthorProvidedOrderWithConfig('after', plugin.name, afters, config, aliasMap);
+      }
+    });
+
+    var aliasArray = Object.keys(aliasMap)
+      .reduce(function(aliases, addonName) {
+        return aliases.concat(aliasMap[addonName].as);
+      }, []);
+
+    var configNames = Object.keys(config).reduce(function(aliasArray, key) {
+      if (aliasArray.indexOf(key) === -1) {
+        aliasArray.push(key);
+      }
+
+      var befores = self._asArray(config[key], 'before');
+
+      befores.forEach(function(name) {
+        if (aliasArray.indexOf(name) === -1) {
+          aliasArray.push(name);
+        }
+      });
+
+      var afters = self._asArray(config[key], 'after');
+
+      afters.forEach(function(name) {
+        if (aliasArray.indexOf(name) === -1) {
+          aliasArray.push(name);
+        }
+      });
+
+      return aliasArray;
+    }, []);
+
+    var unknownConfigKeys = _.difference(configNames, aliasArray);
+
+    if (unknownConfigKeys.length) {
+      this._logUnknownPlugins(unknownConfigKeys, 'config.pipeline.runOrder');
+    }
+
+    return config;
+  },
+
+  _installedPlugins: function(addons, aliasMap) {
+    return Object.keys(addons)
+      .map(function(addonName) {
+        var addon   = addons[addonName];
+        var aliases = aliasMap[addonName].as;
+
+        return aliases.map(function(alias) {
+          var v = addon.createDeployPlugin({ name: alias });
+          return v;
+        });
+      })
+      .reduce(function(plugins, arr) {
+        return plugins.concat(arr);
+      }, []);
+  },
+
+  _applyDisabledConfig: function(plugins, disabledMap) {
+    return plugins.filter(function(plugin) {
+      return !disabledMap[plugin.name];
+    }) ;
+  },
+
+  _applyRunOrderConfig: function(plugins, runOrderMap) {
+    var self          = this;
+    var graph         = new DAG();
+    var sortedPlugins = [];
+
+    try {
+      plugins.forEach(function(plugin) {
+        var before = self._asArray(runOrderMap[plugin.name], 'before');
+        var after = self._asArray(runOrderMap[plugin.name], 'after');
+        graph.addEdges(plugin.name, plugin, before, after);
+      });
+
+      graph.topsort(function (vertex) {
+        sortedPlugins.push(vertex.value);
+      });
+    } catch(err) {
+      if (/cycle detected/.test(err)) {
+        throw new SilentError('your ember-cli-deploy plugins have a circular dependency:' + err.message);
+      } else {
+        throw err;
+      }
+    }
+
+    return sortedPlugins;
+  },
+
+  _isDeployPluginPack: function(addon) {
+    return this._addonHasKeyword(addon, 'ember-cli-deploy-plugin-pack');
+  },
+
+  _isValidDeployPlugin: function(addon) {
+    return this._addonHasKeyword(addon, 'ember-cli-deploy-plugin') && this._addonImplementsDeploymentHooks(addon);
+  },
+
+  _pluginNameFromAddonName: function(addon) {
+    var pluginNameRegex = /^(ember\-cli\-deploy\-)(.*)$/;
+    return addon.name.match(pluginNameRegex)[2];
```

Comment:
on an initial test in a real project this ends up erroring badly

I think you need to guard this or maybe there's something else going on?